### PR TITLE
tikv-ctl: support split region by key

### PR DESCRIPTION
**Pull request by hicqu**
## What have you changed? (mandatory)
Add a new feature for tikv-ctl so that we can split a region by g key.
The new command queries needed information from PD, and then call a tikv rpc. Usage:
```bash
tikv-ctl --pd localhost:12345 split-region -r 2 -k abcde.
```

## What are the type of the changes? (mandatory)
New feature.

## How has this PR been tested? (mandatory)
make dev and test the function in a test cluster.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
It needs doc in docs-cn. But we can add it later.

## Does this PR affect tidb-ansible update? (mandatory)
No.

### CODE REVIEW

File: src/bin/tikv-ctl.rs [hunk 1]
```diff
@@ -1703,3 +1786,42 @@ fn compact_whole_cluster(
...
+            end_number += 255;
+        }
+
+        let digit = (start_number + end_number) / 2;
+
+        if digit == start_number && digit != end_number {
+            end_fill_255 = true;
+            target.push(start[i]);
+        } else if digit > 255 {
+            target[i - 1] += 1; // target[i - 1] must be less than 255.
```
Comment (nolouch) [inline src/bin/tikv-ctl.rs]: `i-1` maybe  -1?

File: src/bin/tikv-ctl.rs [hunk 2]
```diff
@@ -1392,39 +1404,66 @@ fn main() {
...
+                        .required(true)
+                        .takes_value(true)
+                        .help("the target region id")
+                )
+                .arg(
+                    Arg::with_name("key")
+                        .short("k")
+                        .required(true)
+                        .takes_value(true)
+                        .help("the key to split it, in unecoded escaped format")
```
Comment (nolouch) [inline src/bin/tikv-ctl.rs]: Can we use escaped format?  the key usually print with the escaped format.

File: src/bin/tikv-ctl.rs [hunk 3]
```diff
@@ -1656,6 +1695,50 @@ fn dump_snap_meta_file(path: &str) {
...
+        TikvClient::new(channel)
+    };
+
+    let mut req = SplitRegionRequest::new();
+    req.mut_context().set_region_id(region_id);
+    req.mut_context()
+        .set_region_epoch(region.get_region_epoch().clone());
+    req.set_split_key(key.clone());
+
+    tikv_client
```
Comment (nolouch) [inline src/bin/tikv-ctl.rs]: check the response.

**Comment by siddontang** [comment]
can it work now and help me fix the big region problem in the customer environment? 

@nolouch @hicqu

**Comment by hicqu** [comment]
The "RegionNotFound" issue is resolved. I'm going to test it again.

**Comment by hicqu** [comment]
ping @zhangjinpeng1987 @nolouch PTAL.

**Comment by hicqu** [comment]
@zhangjinpeng1987 @nolouch PTAL thanks.

File: src/bin/tikv-ctl.rs [hunk 4]
```diff
@@ -1705,3 +1817,71 @@ fn compact_whole_cluster(
...
+    let mut target = Vec::new();
+    while i < start.len() {
+        let start_number = u16::from(start[i]);
+        let mut end_number = u16::from(end[i]);
+        if end_fill_256 {
+            end_number += 256;
+            end_fill_256 = false;
+        }
+
+        let mut digit = match (start_number & 1, end_number & 1) {
```
Comment (zhangjinpeng1987) [inline src/bin/tikv-ctl.rs]: These lines are unreadable :<

Comment (zhangjinpeng1987) [reply src/bin/tikv-ctl.rs]: It's better to use `(start + end) % 2`.

File: src/bin/tikv-ctl.rs [hunk 5]
```diff
@@ -1705,3 +1820,70 @@ fn compact_whole_cluster(
...
+    let mut end_fill_256 = false;
+    let mut target = Vec::new();
+    while i < start.len() {
+        let start_number = u16::from(start[i]);
+        let mut end_number = u16::from(end[i]);
+        if end_fill_256 {
+            end_number += 256;
+            end_fill_256 = false;
+        }
+
```
Comment (zhangjinpeng1987) [inline src/bin/tikv-ctl.rs]: let sum = start_number + end_number;
let mid = sum / 2;
let shift = sum % 2 != 0;

File: src/bin/tikv-ctl.rs [hunk 6]
```diff
@@ -1705,3 +1820,70 @@ fn compact_whole_cluster(
...
+        (b"ab".to_vec(), b"ac".to_vec(), b"ab".to_vec()),
+        (b"ab".to_vec(), b"ad".to_vec(), b"ac".to_vec()),
+        (
+            b"ab".to_vec(),
+            b"abccc".to_vec(),
+            vec![b'a', b'b', 49, 177, 177],
+        ),
+        (vec![1, 1, 1], vec![2, 1, 1], vec![1, 129, 1]),
+        (vec![1, 200, 1], vec![2, 200, 1], vec![2, 72, 1]),
+        (vec![1, 255, 200], vec![2, 1, 200], vec![2, 0, 200]),
```
Comment (nolouch) [inline src/bin/tikv-ctl.rs]: add a case like left [1,0,0,0,0,1], right [2,255,255,255,255]

**Pull request merged**